ambari-server/src/main/java/org/apache/ambari/server/api/services/stackadvisor/commands/KerberosConfigurationRecommendationCommand.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.ambari.server.controller.internal.AmbariServerConfigurationHandler;
+import org.apache.ambari.server.state.ServiceInfo;
+
+/**
+ * {@link KerberosConfigurationRecommendationCommand} implementation for
+ * Kerberos configuration recommendations.
+ */
+public class KerberosConfigurationRecommendationCommand extends
+    StackAdvisorCommand<RecommendationResponse> {
+
+  public KerberosConfigurationRecommendationCommand(File recommendationsDir,
+                                                    String recommendationsArtifactsLifetime,
+                                                    ServiceInfo.ServiceAdvisorType serviceAdvisorType,
+                                                    int requestId,
+                                                    StackAdvisorRunner saRunner,
+                                                    AmbariMetaInfo metaInfo, AmbariServerConfigurationHandler ambariServerConfigurationHandler) {
+    super(recommendationsDir, recommendationsArtifactsLifetime, serviceAdvisorType, requestId, saRunner, metaInfo, ambariServerConfigurationHandler);
+  }
+
+  @Override
+  protected StackAdvisorCommandType getCommandType() {
+    return StackAdvisorCommandType.RECOMMEND_CONFIGURATIONS_FOR_KERBEROS;
+  }
+
+  @Override
+  protected void validate(StackAdvisorRequest request) throws StackAdvisorException {
+    if (request.getHosts() == null || request.getHosts().isEmpty() || request.getServices() == null
+        || request.getServices().isEmpty()) {
+      throw new StackAdvisorException("Hosts and services must not be empty");
+    }
+  }
+
+  @Override
+  protected RecommendationResponse updateResponse(StackAdvisorRequest request, RecommendationResponse response) {
+    response.getRecommendations().getBlueprint().setHostGroups(processHostGroups(request));
+    response.getRecommendations().getBlueprintClusterBinding().setHostGroups(processHostGroupBindings(request));
+    return response;
+  }
+
+  protected Set<HostGroup> processHostGroups(StackAdvisorRequest request) {
+    Set<HostGroup> resultSet = new HashSet<>();
+    for (Map.Entry<String, Set<String>> componentHost : request.getHostComponents().entrySet()) {
+      String hostGroupName = componentHost.getKey();
+      Set<String> components = componentHost.getValue();
+      if (hostGroupName != null && components != null) {
+        HostGroup hostGroup = new HostGroup();
+        Set<Map<String, String>> componentsSet = new HashSet<>();
+        for (String component : components) {
+          Map<String, String> componentMap = new HashMap<>();
+          componentMap.put("name", component);
+          componentsSet.add(componentMap);
+        }
+        hostGroup.setComponents(componentsSet);
+        hostGroup.setName(hostGroupName);
+        resultSet.add(hostGroup);
+      }
+    }
+    return resultSet;
+  }
+
+  private Set<BindingHostGroup> processHostGroupBindings(StackAdvisorRequest request) {
```

Comment:
Can you doc the private/protected methods in here?

Reply:
This was code copied from another class.. technically I should created an abstract class but some unit test was not happy with something about that.  I will try to figure out what the appropriate javadoc is and fix it in this or in the abstract class I create to provide reusable code.